Sources/NIOCore/AsyncChannelAdapter.swift

```diff
@@ -0,0 +1,514 @@
+extension NIOAsyncChannelWriterHandler {
+    @usableFromInline
+    struct Delegate: @unchecked Sendable, NIOAsyncWriterSinkDelegate {
+        @usableFromInline
+        typealias Element = OutboundOut
+
+        @usableFromInline
+        let loop: EventLoop
+
+        @usableFromInline
+        let handler: NIOAsyncChannelWriterHandler<OutboundOut>
+
+        @inlinable
+        init(handler: NIOAsyncChannelWriterHandler<OutboundOut>) {
+            self.loop = handler.loop
+            self.handler = handler
+        }
+
+        @inlinable
+        func didYield(contentsOf sequence: Deque<OutboundOut>) {
+            self.loop.execute {
+                self.handler._didYield(sequence: sequence)
+            }
+        }
+
+        @inlinable
+        func didTerminate(error: Error?) {
+            // This always called from an async context, so we must loop-hop.
+            self.loop.execute {
+                self.handler._didTerminate(error: error)
+            }
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOInboundChannelStream<InboundIn: Sendable>: Sendable {
+    @usableFromInline
+    typealias Producer = NIOThrowingAsyncSequenceProducer<InboundIn, Error, NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark, NIOAsyncChannelAdapterHandler<InboundIn>.Delegate>
+
+    @usableFromInline let _producer: Producer
+
+    @inlinable
+    init(_ channel: Channel, backpressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark?, closeRatchet: CloseRatchet) throws {
+        channel.eventLoop.preconditionInEventLoop()
+        let handler = NIOAsyncChannelAdapterHandler<InboundIn>(loop: channel.eventLoop, closeRatchet: closeRatchet)
+        let strategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark
+
+        if let userProvided = backpressureStrategy {
+            strategy = userProvided
+        } else {
+            // Default strategy. These numbers are fairly arbitrary, but they line up with the default value of
+            // maxMessagesPerRead.
+            strategy = .init(lowWatermark: 2, highWatermark: 10)
+        }
+
+        let sequence = Producer.makeSequence(backPressureStrategy: strategy, delegate: NIOAsyncChannelAdapterHandler<InboundIn>.Delegate(handler: handler))
+        handler.source = sequence.source
+        try channel.pipeline.syncOperations.addHandler(handler)
+        self._producer = sequence.sequence
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOInboundChannelStream: AsyncSequence {
+    public typealias Element = InboundIn
+
+    public struct AsyncIterator: AsyncIteratorProtocol {
+        @usableFromInline var _iterator: Producer.AsyncIterator
+
+        @inlinable
+        init(_ iterator: Producer.AsyncIterator) {
+            self._iterator = iterator
+        }
+
+        @inlinable public func next() async throws -> Element? {
```

Comment:
Just a drive by we should make this `mutating` before we forget it. I missed to make the `NIOAsyncSequenceProducer.next()` mutating which was mistake but we can correct it here